### ISSUE TITLE
How to apply gamma correction on output of render_torch()?

### ISSUE THREAD

**Issue by mehrab2603** [mitsuba-renderer/mitsuba2#124]
The output of the normal `render()` function can be used to make a `Bitmap` object on which I can call `convert()` to apply gamma correction. How can I do the same for the output of `render_torch()` which is a tensor and cannot be used to create Bitmap?

**Comment by mworchel** [comment]
My guess without having look too much into it: Since you can convert a torch tensor to a numpy array you should be able to use [`this constructor`](https://mitsuba2.readthedocs.io/en/latest/generated/core_api.html#id3) which creates a `Bitmap` from a numpy array (and a pixel format).

**Comment by mehrab2603** [comment]
Does that preserve the computational graph though? I need gradient to flow through the gamma corrected output.

**Comment by mworchel** [comment]
The conversion to a numpy array requires it to be detached from the computational graph, so no. Even the output of the conventional `render()` call has to be converted to a numpy array before it can be used to construct `Bitmap`. So in both cases, the gradients will not propagate through the `Bitmap.convert` method.

**Comment by mehrab2603** [comment]
Too bad then. I also notice that the output of `render()` has values larger than 1. I need the values to be between 0 and 1. Is there a way to have `render()` output values normalized in that range?

**Comment by Speierers** [comment]
To answer your first question, the `Bitmap` class is only implemented on simple `float` (not templated like the rest of the codebase) so operation like gamma correction wouldn't be differentiable in this case. I would recommand to implement this in your python script directly using pytorch or enoki.

Regarding the greater than 1 values in the ouptut image, those are expected as you are using an hdrfilm. You can always post process the output data before computing the loss if it requires values between 0 and 1. (clamping, softmax, ...) Just make sure to do this in a differentiable way (e.g. using pytoch or enoki)

**Comment by mehrab2603** [comment]
@Speierers Does ldrfilm from Mitsuba 1 output values between 0 and 1? If it does, I will try porting it.

**Comment by Speierers** [comment]
Not sure about that, you will have to check yourself
